server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,596 @@
+#! /usr/bin/env python3
+
+import os
+import sys
+import glob
+import tempfile
+
+from pathlib import Path
+from pbench.server import PbenchServerConfig
+from pbench.common.exceptions import BadConfig
+from pbench.common.logger import get_pbench_logger
+from pbench.server.report import Report
+from configparser import ConfigParser, NoSectionError, NoOptionError
+from pbench.server.hierarchy import (
+    Hierarchy,
+    ArchiveHierarchy,
+    ControllerHierarchy,
+    IncomingHierarchy,
+    ResultsHierarchy,
+    UserHierarchy,
+)
+
+
+_NAME_ = "pbench-audit-server"
+
+
+class PbenchMDLogConfig(object):
+    """A simple class to wrap a ConfigParser object in order to query a specific
+    metadata.log file for a given tarball.
+    """
+
+    def __init__(self, cfg_name):
+        self.conf = ConfigParser()
+        self.files = self.conf.read([cfg_name])
+
+    def get(self, *args, **kwargs):
+        return self.conf.get(*args, **kwargs)
+
+
+def verify_valid_controllers(hier, controllers):
+    """Find all the non-directory files at the same level of the controller
+    directories and report them, and find all the normal controller directories.
+    """
+    for controller in controllers:
+        controller_path = Path(controller)
+        if controller_path.is_dir():
+            hier.add_controller(controller_path.name)
+        else:
+            hier.add_bad_controller(controller)
+    return 0
+
```

Comment:
Ideally I still would have liked to see all the separate validation functions defined as methods in the relevant classes to keep the object cohesive and decoupled (between sibling classes and from the main program).

However, I do understand that converting the somewhat chaotic shell script was messy enough that it probably makes more sense to tackle this in stages and not burden this too much with totally "clean coding" guidelines. You've made a lot of improvements in addition to the basic Python conversion, and maybe we should leave it at that for now!